### ISSUE TITLE
Cant freeze graph, to use it in TensorRT

### ISSUE THREAD

**Issue by lubagov** [tensorpack/tensorpack#1126]
Hello!
I have no experience with TensorFlow или TensorPack, but I am trying to start.

I'm trying to convert MaskRCNN solutions for use in NVIDIA TensorRT.
First I need to freeze the graph. I tied this model `COCO-R101FPN-MaskRCNN-BetterParams.npz`. 
And added several lines in the train.py file:
```
predcfg = PredictConfig (
                model = MODEL,
                session_init = get_model_loader (args.load),
                input_names = MODEL.get_inference_tensor_names () [0],
                output_names = ['output / boxes', 'output / scores', 'output / labels'])
ModelExporter (predcfg) .export_compact (pbfile, optimize = False)
```
In fact, I tried many export options, including TensorFlow `checkpoint`. But the result is the same.

Then I convert to uff:
```
python3.5 /usr/lib/python3.5/dist-packages/uff/bin/convert_to_uff.py ./compact_graph.pb -O output/boxes output/scores output/labels
```

I get an error [error.txt](https://github.com/tensorpack/tensorpack/files/3026359/error.txt) In fact, it reports that the graph is not frozen. And the training nodes are not removed from the graph.

At the same time,  Matterport solution works. If I follow the guide https://docs.nvidia.com/deeplearning/sdk/tensorrt-developer-guide/index.html#samplecode2

Or just call `tf.graph_util.convert_variables_to_constants`, and save graph, it works.

If I remove the `tf.graph_util.convert_variables_to_constants` and also remove the ` keras.backend.set_learning_phase (0) ` before exporting, I get the same errors/warnings, in Matterport solution.

Has some thing like `set_learning_phase`, in TensorPack? I try to use `tf.graph_util.remove_training_nodes` but no any effect.

Generally, if someone have some ideas. I try to find a solution to search for cars in real time. I have try many pre-trained models, but in fact, only ResNet101 + FasterCNN on the COCO dataset works well. 
(Shooting occurs a banner by the road, or parking, the CNN  on the COCO dataset, does not work well when viewed from above).

Your pretrained networks, work better than Matterport, I confirm it in my tests. But I need to do it faster. I can see that in solutions in Python, most of the time the mask is calculated on the CPU, if I delete it gets about 4 frames per second - 8 times faster. (there are many cars in the photo, and a mask, is created for all). This is better but not enough. My choice is to use TensorRT to optimize the model (possibly DepStream).

**Comment by ppwwyyxx** [comment]
This is a bug/limitation in nvidia's tensorrt - it cannot convert the graph that tensorpack uses.

**Comment by ppwwyyxx** [comment]
One thing that you can try:
in `basemodel.py`, replace all the 
```
tf.pad(.....)
Conv2D(..., padding='VALID')
```
By 
```
Conv2D(..., padding='SAME')
```

i.e., remove pad, and change valid to same.
This will make TensorRT happier. But it may have other bugs after this one.

This will also produce slightly different results, but probably negligible.

**Comment by lubagov** [comment]
These changes have not changed anything. It seems that I am doing something wrong when exporting the graph.

**Comment by ppwwyyxx** [comment]
> These changes have not changed anything

You should include in your issue precisely what you did and what you observed, like what you did in the original issue.

I would on the other hand conclude that my suggestions were not correctly followed. 
In my test: the following diff does make the error message you've seen disappear:
```diff
--- i/examples/FasterRCNN/basemodel.py
+++ w/examples/FasterRCNN/basemodel.py
@@ -200,12 +200,7 @@ def resnet_fpn_backbone(image, num_blocks):
     with backbone_scope(freeze=freeze_at > 0):
         chan = image.shape[1]
         pad_base = maybe_reverse_pad(2, 3)
-        l = tf.pad(image, tf.stack(
-            [[0, 0], [0, 0],
-             [pad_base[0], pad_base[1] + pad_shape2d[0]],
-             [pad_base[0], pad_base[1] + pad_shape2d[1]]]))
-        l.set_shape([None, chan, None, None])
-        l = Conv2D('conv0', l, 64, 7, strides=2, padding='VALID')
+        l = Conv2D('conv0', image, 64, 7, strides=2, padding='SAME')
         l = tf.pad(l, [[0, 0], [0, 0], maybe_reverse_pad(0, 1), maybe_reverse_pad(0, 1)])
         l = MaxPooling('pool0', l, 3, strides=2, padding='VALID')
     with backbone_scope(freeze=freeze_at > 1):
```

Then I found there is another operation that TensorRT does not support. Another change would make TensorRT produce a UFF:
```diff
diff --git i/examples/FasterRCNN/model_fpn.py w/examples/FasterRCNN/model_fpn.py
index b7c776d8..46b49e26 100644
--- i/examples/FasterRCNN/model_fpn.py
+++ w/examples/FasterRCNN/model_fpn.py
@@ -32,17 +32,13 @@ def fpn_model(features):
     use_gn = cfg.FPN.NORM == 'GN'
 
     def upsample2x(name, x):
-        return FixedUnPooling(
-            name, x, 2, unpool_mat=np.ones((2, 2), dtype='float32'),
-            data_format='channels_first')
-
         # tf.image.resize is, again, not aligned.
-        # with tf.name_scope(name):
-        #     shape2d = tf.shape(x)[2:]
-        #     x = tf.transpose(x, [0, 2, 3, 1])
-        #     x = tf.image.resize_nearest_neighbor(x, shape2d * 2, align_corners=True)
-        #     x = tf.transpose(x, [0, 3, 1, 2])
-        #     return x
+        with tf.name_scope(name):
+            shape2d = tf.shape(x)[2:]
+            x = tf.transpose(x, [0, 2, 3, 1])
+            x = tf.image.resize_nearest_neighbor(x, shape2d * 2, align_corners=True)
+            x = tf.transpose(x, [0, 3, 1, 2])
+            return x
 
     with argscope(Conv2D, data_format='channels_first',
                   activation=tf.identity, use_bias=True,
```

Note that after the two changes, the model require the width & height of input images to be a multiple of 32. It may also produce slightly different results because it uses different operations.

I'm using TensorRT 5.0.2.6 + TF 1.13.1

Closing as this is not a tensorpack issue.

**Comment by lubagov** [comment]
O! Very big, thank you for solution, it generate uff file!
But, `resize_nearest_neighbor`, operation does not supported TensorRT, so I need to replace or implement it. But, a lot, unsupported operations still in graph (if look on warnings). 
In Matterport solution only `resize_nearest_neighbor`, and specific layers, for Mask-RCNN.

**Comment by lubagov** [comment]
I at all has not experience with this, just trying. In Keras has option `keras.backend.set_learning_phase` and it working.
If i set `keras.backend.set_learning_phase (1)` then i get a lot warning for unsupported operations.
If i set `keras.backend.set_learning_phase (0)` all warnings exclude `resize_nearest_neighbor` (and custom layers) disappear. Don't know how it working, and what to do this option, but i believe, it could somehow work, will try to found :-)

**Comment by ppwwyyxx** [comment]
Keras' learning phase has nothing to do with the issue here. When you export a graph in tensorpack with PredictConfig, it already uses the testing phase.

The issue is tensorpack's implementation uses more types of ops in tensorflow, some of those unsupported by tensorrt.
